### ISSUE TITLE
Accessing variable as object, but in some cases it could be a string

### ISSUE THREAD

**Issue by GChernikov** [silverstripe/silverstripe-restfulserver#11]
restfulserver/code/RestfulServer.php

There is a method:
protected function updateDataObject($obj, $formatter) 

which can return $obj (type DataObject) or string 'No Content'. And then in places where this method called returning value is accessing with assumption, that there could be only DataObject. Do a search for:
$obj->class/$obj->ID
in putHandler / postHandler methods.

**Comment by robbieaverill** [comment]
Without breaking the API (we probably should) we can update the doc blocks and add some edge case handling for when a string is returned.

Having a string here is actually not compatible with the restful server code - in terms of this example:

```php
        // ...
        /** @var DataObject|string */
        $obj = $this->updateDataObject($obj, $reqFormatter);

        return $responseFormatter->convertDataObject($obj);
```

In this case `$responseFormatter` is a `DataFormatter` subclass, and `convertDataObject()` has a type hinted `DataObjectInterface` for its argument, so we can't actually pass a string into it at all.

We will need to add some extra logic to handle the string return earlier.

**Comment by robbieaverill** [comment]
This bug can be reproduced with this test:

```php
public function testPostWithoutBodyReturnsNoContent()
{
    $_SERVER['PHP_AUTH_USER'] = 'editor@test.com';
    $_SERVER['PHP_AUTH_PW'] = 'editor';

    $url = '/api/v1/RestfulServerTest_Comment';
    $response = Director::test($url, null, null, 'POST');

    $this->assertEquals('No Content', $response->getBody());

    unset($_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW']);
}
```

```
1) RestfulServerTest::testPostWithoutBodyReturnsNoContent
Trying to get property of non-object

/Users/robbieaverill/dev/cwp17/restfulserver/code/RestfulServer.php:488
/Users/robbieaverill/dev/cwp17/restfulserver/code/RestfulServer.php:130
/Users/robbieaverill/dev/cwp17/framework/control/RequestHandler.php:288
/Users/robbieaverill/dev/cwp17/framework/control/Controller.php:202
/Users/robbieaverill/dev/cwp17/framework/control/RequestHandler.php:200
/Users/robbieaverill/dev/cwp17/framework/control/Controller.php:158
/Users/robbieaverill/dev/cwp17/framework/control/Director.php:385
/Users/robbieaverill/dev/cwp17/framework/control/Director.php:319
/Users/robbieaverill/dev/cwp17/restfulserver/tests/unit/RestfulServerTest.php:202
/Users/robbieaverill/dev/cwp17/vendor/phpunit/phpunit/composer/bin/phpunit:63
```

**Comment by robbieaverill** [comment]
Fixed with #40
